DS-2600: Choose how the filename of a bitstream is generated when is requested for download

FacundoAdorno opened the pull request:
When a bitstream is downloaded, the original bitstream's name ('dc.title' metadata) is used to set the filename.
This new feature lets set, in four different ways, how the filename is generated, by defining the following configuration options:
- TITLE: set the filename with the bitstream original name (Bitstream.getName()).
- DESCRIPTION: set the filename with the bitstream description (Bitstream.getDescription())
- PARAMETER: set the filename with the parameter of the url. P.e.: in &lt;dspace.url&gt;/handle/&lt;prefix&gt;/&lt;sufix&gt;/&lt;filename_in_url&gt;?&lt;sequence&gt;..., the filename value is &lt;filename_in_url&gt;.
- HANDLE: set the filename using the pattern &lt;handlePrefix&gt;_&lt;handleSufix&gt;_&lt;bitstream_sequence&gt;. P.e.: 123456789_325_2.

If none of the previous options can be applied, then the generic name 'bitstream' is used.

Once the download request is processed, the filename will be set in the element "Content-Disposition" of the HTTP response header. To do this, a little modification was made to ensure that the "Content-Disposition" element be always generated.

FacundoAdorno committed:
For each bitstream download, the element "Content-Disposition" of the HTTP response header always will have some of the following attributes: 'inline' or 'attachment'.

FacundoAdorno committed:
Modification to generate the filename, of a downloaded file, from four differents sources: bitstream title, bitstream description, item handle, parameter name in URL.

tdonohue commented:
Thanks for the PR, @FacundoAdorno .  At a glance, the code looks decent.

Though, there are some areas that could use minor tweaks -- e.g. replacing some of those "if"s with "else if" calls (or a switch statement), just so that each if statement doesn't need to processed after a single one matches.

We'll also need a few developers to test this out on our end to consider for possible inclusion in 6.0. Anyone else watching this PR, please let us know if you can help test this out!

FacundoAdorno commented:
Hi Tim, thanks for the reply! I'll try to clean up the code and make it clearer.

FacundoAdorno committed:
Clean up the code of the filename generation

The code of the filename generation feature, included in the
previous commit (7ad3599e), was cleaned up to make it clearer.

FacundoAdorno commented:
I've made ​​a cleaning up of the code . Waiting for volunteers! Regards

FacundoAdorno commented:
Travis reports the below error. I don't think that it can be something related to my code. Could be possible re run Travis?
```
Running org.dspace.content.packager.ITDSpaceAIP
Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 34.956 sec - in org.dspace.content.packager.ITDSpaceAIP
Running org.dspace.content.ITCommunityCollection
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 69.425 sec <<< FAILURE! - in org.dspace.content.ITCommunityCollection
testCountItems(org.dspace.content.ITCommunityCollection)  Time elapsed: 66.76 sec  <<< FAILURE!
junit.framework.AssertionFailedError: 95-percentile of org.dspace.content.ITCommunityCollection.testCountItems exceeded the requirement of 2000 ms, measured 2145 ms
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.databene.contiperf.report.ReportContext.fail(ReportContext.java:68)
	at org.databene.contiperf.PerformanceTracker.checkRequirements(PerformanceTracker.java:180)
	at org.databene.contiperf.PerformanceTracker.stopTracking(PerformanceTracker.java:127)
	at org.databene.contiperf.junit.PerfTestStatement.evaluate(PerfTestStatement.java:82)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Running org.dspace.content.ITMetadata
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.345 sec - in org.dspace.content.ITMetadata
Results :
Failed tests: 
  ITCommunityCollection.testCountItems 95-percentile of org.dspace.content.ITCommunityCollection.testCountItems exceeded the requirement of 2000 ms, measured 2145 ms
```
Complete [stacktrace](http://pastebin.com/phTFqNJK).

FacundoAdorno commented:
Hello! The Travis build has been passed successfully!

tdonohue commented:
Closing as this PR must be recreated on `dspace-6_x` branch. XMLUI has been removed from `master`. See this mailing list notice: https://groups.google.com/d/msg/dspace-devel/aEFqSn46ryM/iPRr_IMLCwAJ
